[CASSANDRA-20084][4.1] Fix the case when WaitQueue.Signal.awaitUninterruptibly() may stuck forever if the invoking thread is interrupted

netudima opened the pull request:
Fix the case when WaitQueue.Signal.awaitUninterruptibly() may stuck forever if the invoking thread is interrupted

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20084

Commit:
Fix the case when WaitQueue.Signal.awaitUninterruptibly() may stuck forever if the invoking thread is interrupted

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20084

Commit:
Changed the implementation based on the review comments, added a test for Signal.awaitUninterruptibly(long, TimeUnit) too

Patch by Dmitry Konstantinov; reviewed by Benedict Elliott Smith for CASSANDRA-20084

Commit:
Return back to the original version with explicit awaitUninterruptibly and add the same logic for awaitUntilUninterruptibly

Patch by Dmitry Konstantinov; reviewed by Benedict Elliott Smith for CASSANDRA-20084

Commit:
inline inner check logic into awaitUninterruptibly/awaitUntilUninterruptibly to make them simpler

Patch by Dmitry Konstantinov; reviewed by Benedict Elliott Smith for CASSANDRA-20084

Commit:
checkAndClear is needed in awaitUninterruptibly because it is public and can be overridden in a sub-class to do extra work

Patch by Dmitry Konstantinov; reviewed by Benedict Elliott Smith for CASSANDRA-20084

Commit:
use isSet in awaitUntil too

Patch by Dmitry Konstantinov; reviewed by Benedict Elliott Smith for CASSANDRA-20084

belliottsmith commented:
LGTM, though I haven't had time to be exhaustive about other possible codepaths that might be affected, a quick check looks like they're covered, and I'm confident this patch makes things better than they were.

jmckenzie-dev commented on src/java/org/apache/cassandra/utils/concurrent/WaitQueue.java:
```diff
@@ -74,7 +74,7 @@
  * <p>5. This scheme is not fair</p>
  * <p>6. Only the thread that calls register() may call await()</p>
- *
+ * <p>6. A signal can be cancelled while it is within await() if the invoking thread is interrupted</p>
```
nit: think this should be point 7 shouldn't it?

jmckenzie-dev commented:
+1 from me as well w/one nit about numbering of new item in the class comment. Dug around a bit in adjacent callers and impact and I can't come up with a plausible argument on how this could make anything regress.

Commit:
address review comments: fix numbering for the added comment

Patch by Dmitry Konstantinov; reviewed by Benedict Elliott Smith,Josh McKenzie for CASSANDRA-20084
